src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,972 @@
+            var expressionDecl = expression?.Parent?.Parent;
+
+            localDeclaration = null;
+
+            if (!syntaxFacts.IsVariableDeclarator(expressionDecl))
+            {
+                varDeclName = null;
+                return false;
+            }
+
+            localDeclaration = GetLocalDeclarationFromDeclarator(expressionDecl);
+            if (localDeclaration is null)
+            {
+                varDeclName = null;
+                return false;
+            }
+
+            // TODO: handle in the future
+            if (syntaxFacts.GetVariablesOfLocalDeclarationStatement(localDeclaration).Count > 1)
+            {
+                varDeclName = null;
+                localDeclaration = null;
+                return false;
+            }
+
+            varDeclName = syntaxFacts.GetIdentifierOfVariableDeclarator(expressionDecl).ValueText;
+            return true;
+        }
+
+        /// <summary>
+        /// Locates all the call sites of the method that introduced the parameter
+        /// </summary>
+        protected static async Task<ImmutableDictionary<Document, List<SyntaxNode>>> FindCallSitesAsync(
+            Document document, IMethodSymbol methodSymbol, CancellationToken cancellationToken)
+        {
+            var methodCallSitesBuilder = ImmutableDictionary.CreateBuilder<Document, List<SyntaxNode>>();
+            var progress = new StreamingProgressCollector();
+            await SymbolFinder.FindReferencesAsync(
+                methodSymbol, document.Project.Solution, progress,
+                documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
+            var referencedSymbols = progress.GetReferencedSymbols();
+
+            // Ordering by descending to sort invocations by starting span to account for nested invocations
+            var referencedLocations = referencedSymbols.SelectMany(referencedSymbol => referencedSymbol.Locations)
+                .Distinct().Where(reference => !reference.IsImplicit)
+                .OrderByDescending(reference => reference.Location.SourceSpan.Start);
+
+            // Adding the original document to ensure that it will be seen again when processing the call sites
+            // in order to update the original expression and containing method.
+            methodCallSitesBuilder.Add(document, new List<SyntaxNode>());
+
+            foreach (var refLocation in referencedLocations)
+            {
+                // Does not support cross-language references currently
+                if (refLocation.Document.Project.Language == document.Project.Language)
+                {
+                    var reference = refLocation.Location.FindNode(cancellationToken).GetRequiredParent();
+                    while (reference is not (TObjectCreationExpressionSyntax or TInvocationExpressionSyntax))
+                    {
+                        reference = reference.GetRequiredParent();
```

Comment:
this is walking up too high.  you could find a random reference to some random node that's in a lambda in an argument that is eventually in an invocation, and this would match.  That's not ok.  You need to check for the exact shapes we're talking about.  i.e. the direct parent is an object creation, or hte invocation is of the form `Foo(...)` or `x.Foo(...)` or `x?.Foo(...)` 

Or, put another way, you can have this code.  However, once you get the creation/invocation expression, you need to validate that the reference node is the correct child of it (i.e. the name of hte invocation).

Reply:
I think that, in the case where the reference is not already either an ObjectCreationExpressionSyntax or InvocationExpressionSyntax, then the direct parent would be. So I'm not sure I need a while loop at all